pkg/fdsrv/fdsrv.go

```diff
@@ -0,0 +1,212 @@
+	if err != nil {
+		return false, err
+	}
+	query := string(buf[:n])
+	if query != fds.nonce {
+		io.WriteString(uc, "BAD NONCE")
+		return false, nil
+	}
+	oob := syscall.UnixRights(fds.dupedFd)
+	good := []byte("GOOD NONCE")
+	goodn, oobn, err := uc.WriteMsgUnix(good, oob, nil)
+	if err != nil {
+		return false, err
+	}
+	if goodn != len(good) || oobn != len(oob) {
+		return false, errors.New("truncated write")
+	}
+	return true, nil
+}
+
+func NewFdServer(fd int, nonce string,
+	options ...func(*FdServer) error) (*FdServer, error) {
+	var err error
+	fds := &FdServer{}
+
+	if len(nonce) == 0 {
+		return nil, errors.New("nonce must not be empty")
+	}
+	fds.nonce = nonce
+
+	for _, op := range options {
+		if err := op(fds); err != nil {
+			return nil, err
+		}
+	}
+
+	// An empty addr tells Linux to "autobind" to an available path in the
+	// abstract unix domain socket namespace
+	ua, err := net.ResolveUnixAddr("unix", "")
+	if err != nil {
+		return nil, err
+	}
+	fds.listener, err = net.ListenUnix("unix", ua)
+	if err != nil {
+		return nil, err
+	}
+
+	// Caller could close the file while we are running.  Keep our own copy.
+	fds.dupedFd, err = syscall.Dup(int(fd))
+	if err != nil {
+		fds.listener.Close()
+		return nil, err
+	}
+
+	return fds, nil
+}
+
+func WithTimeout(timeout time.Duration) func(*FdServer) error {
+	return func(fds *FdServer) error {
+		fds.timeout = timeout
+		return nil
+	}
+}
+
+func WithServeOnce() func(*FdServer) error {
+	return func(fds *FdServer) error {
+		fds.serveOnce = true
+		return nil
+	}
+}
+
+func (fds *FdServer) UdsPath() string {
+	return fds.listener.Addr().String()
+}
+
+func (fds *FdServer) Nonce() string {
+	return fds.nonce
+}
+
+func (fds *FdServer) Destroy() {
+	fds.listener.Close()
+	syscall.Close(fds.dupedFd)
+}
+
+func (fds *FdServer) Serve() error {
+	var deadline time.Time
+	if fds.timeout != 0 {
+		deadline = time.Now().Add(fds.timeout)
+	}
+	fds.listener.SetDeadline(deadline)
+	for {
+		conn, err := fds.listener.AcceptUnix()
+		if err != nil {
+			return err
+		}
+		conn.SetDeadline(deadline)
+		succeeded, err := fds.handleConnection(conn)
+		if err != nil {
+			return err
+		}
+		if succeeded && fds.serveOnce {
+			break
+		}
+	}
+	return nil
+}
+
+// Client method
+func GetSharedFd(udsPath, nonce string) (int, error) {
+	ua, err := net.ResolveUnixAddr("unix", udsPath)
+	if err != nil {
+		return 0, err
+	}
+	uc, err := net.DialUnix("unix", nil, ua)
+	if err != nil {
+		return 0, err
+	}
+	var nonceb = []byte(nonce)
+
+	n, err := uc.Write(nonceb)
+	if err != nil {
+		return 0, err
+	}
+	// If you don't send at least a byte, the server won't recvmsg.  This
+	// is a Linux UDS SOCK_STREAM thing.  It's possible that nonce is "",
+	// and len(nonceb) == 0.  We could check earlier for an empty string,
+	// but test code uses this method to make sure the server can handle an
+	// empty nonce (BadEmptyNonce).
+	if n == 0 {
+		return 0, errors.New("do not send an empty nonce")
```

Comment:
This is where having the variable is handy, it enforces the same value errors.New("nonce must not be empty")

Reply:
there's actually a subtle difference in the two cases, especially if you go around grepping for errstrs, but it's not worth having its own variable.  =)